Driver probing sequence - step1 - factorize clk_dt code

etienne-lms opened the pull request:
This P-R implements the first commits from RFC P-R #4949 that moves code from **core/drivers/clk/clk_dt.c** to a newly created source file **core/kernel/dt_driver.c** to use the sequences and resources for other drivers than clocks.

jforissier commented on core/include/kernel/dt_driver.h:
```diff
@@ -0,0 +1,63 @@
+ * @nodeoffset: Node offset of device referenced in the FDT
+ * @type: One of DT_DRIVER_* or DT_DRIVER_NOTYPE.
+ * @provider_cells: Cells count in the FDT used by the driver's references
+ * @get_of_device: Callback to get driver's device ref from phandle data
```
s/Callback/Function/

etienne-lms replied:
ok

jforissier commented on core/include/kernel/dt_driver.h:
```diff
@@ -60,4 +61,39 @@ struct dt_driver_provider {
+ *
+ * @get_of_device callback return a void *. Driver provider is expected to
+ * include a shim helper to cast to device reference into provider driver
+ * target structure reference (i.e. (struct clk *) for clock devices.
```
Missing closing parenthese.
`i.e.` means "that is", not "for example" which would be better in this context. Do you mean `e.g.`?

etienne-lms replied:
yes, i'll fix. thanks for the info :)

jforissier commented on core/include/kernel/dt_driver.h:
```diff
@@ -60,4 +61,39 @@ struct dt_driver_provider {
+ *
+ * @fdt: Device tree to work on
+ * @nodeoffset: Node offset of the clock
+ * @get_of_device: Callback to match the devicetree with a device instance
```
s/Callback/Function/

etienne-lms replied:
ok

jforissier commented on core/include/kernel/dt_driver.h:
```diff
@@ -60,4 +61,39 @@ struct dt_driver_provider {
+ * @data: Data which will be passed to the @get_of_device callback
+ * @type: Driver type
+ *
+ * @get_of_device callback return a void *. Driver provider is expected to
```
s/callback//, returns

etienne-lms replied:
ok

jforissier commented on core/include/kernel/dt.h:
```diff
@@ -66,11 +67,35 @@ enum dt_driver_type {
 	enum dt_driver_type type;
 	const struct dt_device_match *match_table; /* null-terminated */
 	const void *driver;
+	TEE_Result (*probe)(const void *fdt, int node, const void *compat_data);
```
Why not use the typedef?

etienne-lms replied:
true, thanks. I'll fix (actaully it's fixed by a later commit in the series).

jforissier commented on core/drivers/clk/clk_dt.c:
```diff
@@ -13,18 +13,6 @@
 #include <libfdt.h>
 #include <stddef.h>
 
-struct clk *clk_dt_get_by_name(const void *fdt, int nodeoffset,
```
> `Move clk_dt_get_by_name() right below clk_dt_get_by_idx().`

Sure, but why? ;-)

etienne-lms replied:
well... true it looks a bit useless.
The change makes th function easier to read as it calls `clk_dt_get_by_idx()` that is rigth below.

I can discard this change.

jforissier reviewed (commented):
I'm OK with this but have a few cosmetic comments below.

etienne-lms commented:
Comments addressed.

etienne-lms commented:
CI for xtest on QEMUv8 fails with the following trace:
```
This system (build) is not supported. See file INSTALL for details.
make[2]: *** [package/pkg-generic.mk:260: /root/optee_repo_qemu_v8/out-br/build/host-libopenssl-1.1.1l/.stamp_configured] Error 1
make[1]: *** [Makefile:23: _all] Error 2
make[1]: Leaving directory '/root/optee_repo_qemu_v8/out-br'
make: *** [common.mk:324: buildroot] Error 2
```
It does not seem related to this P-R.
Did I miss something?

jforissier commented:
@etienne-lms 

> CI for xtest on QEMUv8 fails with the following trace:
[...]
> It does not seem related to this P-R. Did I miss something?

Yes, https://github.com/OP-TEE/optee_os/pull/4964 :wink:

jforissier reviewed (commented):
`Acked-by: Jerome Forissier <jerome@forissier.org>`

jforissier commented:
> may I squash the fixup commits?

yes please

etienne-lms commented:
fixup commits squashed

jforissier commented on core/include/kernel/dt_driver.h:
```diff
@@ -0,0 +1,63 @@
+#include <sys/queue.h>
+
+/**
+ * struct dt_driver_phandle_args - Devicetree clock args
```
clock?

jforissier commented on core/include/kernel/dt_driver.h:
```diff
@@ -60,4 +61,39 @@ struct dt_driver_provider {
+ * dt_driver_register_provider - Register a driver provider
+ *
+ * @fdt: Device tree to work on
+ * @nodeoffset: Node offset of the clock
```
clock?

jforissier commented on core/include/kernel/dt_driver.h:
```diff
@@ -79,6 +79,21 @@ TEE_Result dt_driver_register_provider(const void *fdt, int nodeoffset,
+/*
+ * Return a device instance based on a property name and FDT information
+ *
+ * @prop_name: DT property name, i.e. "clocks" for clock resources
```
"for example" or "e.g."

jforissier commented on core/include/kernel/dt_driver.h:
```diff
@@ -79,6 +79,21 @@ TEE_Result dt_driver_register_provider(const void *fdt, int nodeoffset,
+ * @nodeoffset: node offset in the FDT
+ * @prop_idx: index of the phandle data in the property
+ *
+ * Return a device opaque reference, i.e. a struct clk pointer for a clock
```
"for example" or "e.g."

jforissier commented on core/include/kernel/dt.h:
```diff
@@ -66,11 +67,35 @@ enum dt_driver_type {
+/*
+ * dt_driver_probe_func - Callback probe function for a driver.
+ *
+ * @fdt; FDT base address
```
s/;/:/

etienne-lms commented:
thanks for the findings (and apologies for what you already commented on)

etienne-lms commented:
comments addressed

jforissier commented:
`Acked-by: Jerome Forissier <jerome@forissier.org>`

etienne-lms commented:
@clementleger, since these changes hacks into the clock drivers you posted and maintain, could you ack (or comment) this P-R?
Note: if that helps, I can add myself as 2nd maintainer for the clock drivers (in [MAINTAINERS](https://github.com/OP-TEE/optee_os/blob/48bc45098fa664b1e7c018ea856fe040a8d9817e/MAINTAINERS#L161-L164)).

In the meantime, i'll squash the fixup commits.

clementleger commented on core/include/kernel/dt_driver.h:
```diff
@@ -0,0 +1,116 @@
+				       void *data, enum dt_driver_type type);
+
+/*
+ * Return a device instance based on a property name and FDT information
```
Somtime the function name is present, sometimes not. Maybe add function name here ?

etienne-lms replied:
right, i'll add it where there is a detailed inline description of the function (the prototypes you pointed out).

clementleger commented on core/include/kernel/dt_driver.h:
```diff
@@ -0,0 +1,116 @@
+unsigned int dt_driver_provider_cells(struct dt_driver_provider *prv);
+
+/*
+ * Probe matching driver to create a device from a FDT node
```
ditto

clementleger commented on core/kernel/dt_driver.c:
```diff
@@ -0,0 +1,256 @@
+	default:
+		panic();
+	}
+
```
I think you could use the following to avoid duplicating `fdt_getprop` calls laters:

```
const char *cells_name = NULL;
switch (type) {
case DT_DRIVER_CLK:
	cells_name = "#clock-cells";
	break;
default:
	panic();
}
c = fdt_getprop(fdt, nodeoffset, cells_name, &len);

```

etienne-lms replied:
agree

clementleger reviewed (commented):
Looks good to me apart some minor points.

`Acked-by: Clément Léger <clement.leger@bootlin.com>`

etienne-lms commented on core/kernel/dt_driver.c:
```diff
@@ -0,0 +1,256 @@
+#include <kernel/dt_driver.h>
+#include <libfdt.h>
+#include <malloc.h>
+#include <sys/queue.h>
```
still missing! will fix

etienne-lms replied:
done

etienne-lms commented:
All good to me.
Squashed fixup commits (+add `#include <tee_api_types.h> in _dt_driver.h_) 
Applied review tags (note I changed @clementleger tag s/é/e/ to comply with checkpatch expectations)

jforissier commented:
> Applied review tags (note I changed @clementleger tag s/é/e/ to comply with checkpatch expectations)

Fine with me if he's fine too, but note that we have a bunch of commit from him with the accented characters already.

clementleger commented:
> Fine with me if he's fine too, but note that we have a bunch of commit from him with the accented characters already.

Yes, many commits already contains my name with accent. Maybe you can ignore the checkpatch warning on this one ?

etienne-lms commented:
forced pushed with @clementleger review tag fixed.

Checkpatch will complain (trace below, for the record, strangely hits Clément but not Léger)
```
WARNING: 'ment' may be misspelled - perhaps 'meant'?
#16: 
Acked-by: Clément Léger <clement.x@x.x>
              ^^^^
```

etienne-lms commented:
many thanks for the feedback and help.
